### PR TITLE
[entropy_src,sw] Make KAT test more robust

### PR DESCRIPTION
**Pull request by martin-lueker**
The entropy_src KAT test's use of non-blocking data reads is important. However it leaves the test vulnerable to false failures due to timing
delays within the IP.   This commit issues a number of retry attempts
whenever the ENTROPY_DATA register is read.

Signed-off-by: Martin Lueker-Boden <martin.lueker-boden@wdc.com>

### CODE REVIEW

File: sw/device/tests/entropy_src_kat_test.c [hunk 1]
```diff
@@ -88,10 +88,19 @@ static void flush_sha3_conditioner(dif_entropy_src_t *entropy_src) {
...
+  dif_result_t op_result;
+
   // Read (and discard) the resulting seed.
   uint32_t got[kEntropyFifoBufferSize];
   for (size_t i = 0; i < ARRAYSIZE(got); ++i) {
-    CHECK_DIF_OK(dif_entropy_src_non_blocking_read(entropy_src, &got[i]));
+    op_result = dif_entropy_src_non_blocking_read(entropy_src, &got[i]);
+    if (op_result == kDifUnavailable) {
+      fail_count++;
+      CHECK(fail_count < KAT_TEST_TIMEOUT_ATTEMPTS);
```
Comment (moidx) [inline sw/device/tests/entropy_src_kat_test.c:100]: s/`KAT_TEST_TIMEOUT_ATTEMPTS`/`kKatTestTimeoutAttempts`/ We recommend using enum constants instead to improve debuggability. See how `kEntropyFifoBufferSize` is defined in this test case.

**Review by moidx — APPROVED** [review]

**Comment by andreaskurth** [comment]
Thanks @martin-lueker :+1: If this PR is done from your side, I can implement @moidx's feedback and get this merged? Then it would be off your table

**Comment by martin-lueker** [comment]
> Thanks @martin-lueker 👍 If this PR is done from your side, I can implement @moidx's feedback and get this merged? Then it would be off your table

Hi Andreas,  Sure if you like.  In my opinion though, it has not been a high priority (which is why it has collected some dust). I threw this one up when I was barking up the wrong tree for a top-level CI error at some point.   May be a "nice-to-have" for M2

**Comment by vogelpi** [comment]
Thanks for the feedback @martin-lueker , I'll take care of rebasing this and implementing the minor changes. It's always good to get things out of the way if possible.

**Comment by vogelpi** [comment]
I've now implemented the remaining feedback points and afterwards also rebased this on current master. Now running the chip-level tests in parallel to CI.

**Comment by vogelpi** [comment]
The top-level test passed with Verilator, Xcelium and VCS. We can merge this once CI ran through.

File: sw/device/tests/entropy_src_kat_test.c [hunk 2]
```diff
@@ -88,10 +88,19 @@ static void flush_sha3_conditioner(dif_entropy_src_t *entropy_src) {
...
   stop_sha3_conditioner(entropy_src);
 
+  int fail_count = 0;
+  dif_result_t op_result;
+
   // Read (and discard) the resulting seed.
   uint32_t got[kEntropyFifoBufferSize];
   for (size_t i = 0; i < ARRAYSIZE(got); ++i) {
-    CHECK_DIF_OK(dif_entropy_src_non_blocking_read(entropy_src, &got[i]));
+    op_result = dif_entropy_src_non_blocking_read(entropy_src, &got[i]);
```
[the remark this answers is not in the document]
Comment (vogelpi) [reply sw/device/tests/entropy_src_kat_test.c]: Done.

File: sw/device/tests/entropy_src_kat_test.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/tests/entropy_src_kat_test.c]: Done.

**Pull request merged**